modules/PowermeterBSM_WS36A/main/powermeterImpl.cpp

```diff
@@ -0,0 +1,171 @@
+
+    // Wait for the signature to finish.
+    while (read_register<uint16_t>(SIGNATURE_STATUS_START) == SIGNATURE_STATUS_DONE) {
+    }
+    write_register(SIGNATURE_STATUS_START, SIGNATURE_STATUS_DONE);
+
+    // Read the start signature.
+    const auto response = read_register<std::string>(SIGNATURE_START);
+    EVLOG_info << "Started the transaction with " << response;
+
+    return {TransactionRequestStatus::OK};
+}
+
+TransactionStopResponse powermeterImpl::handle_stop_transaction_impl(const std::string& _) {
+    EVLOG_info << "Stopping transaction";
+
+    // Wait for the signature to finish.
+    while (read_register<uint16_t>(SIGNATURE_STATUS_STOP) == SIGNATURE_STATUS_DONE) {
+    }
+    write_register(SIGNATURE_STATUS_STOP, SIGNATURE_STATUS_DONE);
+
+    // Read the signature.
+    const auto response = read_register<std::string>(SIGNATURE_STOP);
+    EVLOG_info << "Stopped the transaction with " << response;
+
+    return {TransactionRequestStatus::OK, response};
+}
+
+TransactionStartResponse powermeterImpl::handle_start_transaction(TransactionReq& value) {
+    try {
+        return handle_start_transaction_impl(value);
+    } catch (const std::exception& _ex) {
+        EVLOG_error << "Failed to start the transaction: " << _ex.what();
+        return {TransactionRequestStatus::OK, _ex.what()};
+    }
+}
+
+TransactionStopResponse powermeterImpl::handle_stop_transaction(std::string& transaction_id) {
+    try {
+        return handle_stop_transaction_impl(transaction_id);
+    } catch (const std::exception& _ex) {
+        EVLOG_error << "Failed to stop the transaction: " << _ex.what();
+        return {TransactionRequestStatus::OK, {}, _ex.what()};
+    }
+}
+
+template <typename Output> Output powermeterImpl::read_register(const Register& _register) const {
+    const auto read_register_impl = [this](const Register& _register) -> types::serial_comm_hub_requests::Result {
+        switch (_register.type) {
+        case RegisterType::HOLDING: {
+            return mod->r_serial_comm_hub->call_modbus_read_holding_registers(
+                config.powermeter_device_id, _register.start_register - config.modbus_base_address,
+                _register.num_registers);
+        }
+        case RegisterType::INPUT: {
+            return mod->r_serial_comm_hub->call_modbus_read_input_registers(
+                config.powermeter_device_id, _register.start_register - config.modbus_base_address,
+                _register.num_registers);
+        }
+        default:
+            return {};
+        }
+    };
+
+    const auto response = read_register_impl(_register);
+    if (response.status_code != StatusCodeEnum::Success)
+        throw std::runtime_error("Failed to send the message");
+    return deserialize<Output>(response);
+}
+
+template <typename Input> void powermeterImpl::write_register(const Register& register_data, const Input& data) const {
+
+    // Serialize the data into a vector.
+    auto ser_data = serialize(data);
```

Comment:
this makes me think that `serialize` should really be in a namespace that makes it clear that it has to do with modbus. That has to be done in #442 though.

Reply:
don't really think that this is necessary since the whole function is about calling the serial-comm-hubs' modbus methods but will do